lib/kubeclient/common.rb

```diff
@@ -286,6 +308,46 @@ def get_entities(entity_type, klass, resource_name, options = {})
       Kubeclient::Common::EntityList.new(entity_type, resource_version, collection)
     end
 
+    def get_entities_async(klass, resource_name, options = {})
+      params = {}
+      SEARCH_ARGUMENTS.each { |k, v| params[k] = options[v] if options[v] }
+
+      ns_prefix = build_namespace_prefix(options[:namespace])
+
+      resource_version = nil
+      entity_streamer = Json::Streamer::JsonStreamer.new
+      resource_version_streamer = Json::Streamer::JsonStreamer.new
+
+      resource_version_streamer.get(key: 'resourceVersion') do |value|
+        resource_version = value.to_i unless resource_version
+      end
+
+      # result['items'] might be nil due to https://github.com/kubernetes/kubernetes/issues/13096
+      entity_streamer.get(nesting_level: 2, yield_values: false) do |object|
+        entity = new_entity(object, klass)
+        yield(entity, resource_version)
+      end
+
+      block = proc do |response|
+        if response.instance_of?(Net::HTTPNotFound)
+          raise ResourceNotFoundError.new(404,
```

Comment:
I believe RestClient::Resource raises for ANY non-2xx status, not just 404.  Should have similar logic here.
(Consider 5xx errors, or  kubernetes saying 418 I'm a teapot.  Obviously we must support that!)
Also, what about 3xx redirects?

I'd love if we can to converge on just 1 way to call RestClient, either Resource or .execute, for both flows.

Reply:
@cben not sure how to handle 3xx redirects. easiest solution is to simply return ResourceNotFound for any non 2xx status

Reply:
you're now checking both NotFound here and any error in `do_rest_request`.
can you check in one place, e.g. using
`error_klass = e.http_code == 404 ? ResourceNotFoundError : HttpError`

Reply:
Default redirect/error handling is implemented by `Response`, specifically in https://github.com/rest-client/rest-client/blob/v2.0.2/lib/restclient/abstract_response.rb
Perhaps we might want to use it for all non-2xx statuses, like done here:
https://github.com/rest-client/rest-client/blob/v2.0.2/lib/restclient/request.rb#L802-L803

I'm also OK with raising on any non-2xx.

BTW, another processing we're skipping is Content-encoding.  What happens when k8s will enable compression?
(Note `Request.decode` is gone in upcoming rest-client 2.1: https://github.com/rest-client/rest-client/pull/597)

Reply:
good question; it looks like RestClient automatically handles decompression: https://github.com/rest-client/rest-client/blob/f450a0f086f1cd1049abbef2a2c66166a1a9ba71/spec/integration/httpbin_spec.rb#L89